cmd/gogrep/main.go

```diff
@@ -21,44 +23,61 @@ func main() {
 	var before int
 	flag.IntVar(&before, "B", 0, "Shows number of lines before the Match.")
 
+	var recursive bool
+	flag.BoolVar(&recursive, "r", false, "Searches every file inside a directory.")
+
 	flag.Parse()
 
-	var fileContents []string
 	pattern := flag.Arg(0)
-	listOfFiles := iofile.ListFilesInDir(flag.Arg(1))
-	for i := 0; i < len(listOfFiles); i++ {
-		fileContents, _ = iofile.ReadFile(listOfFiles[i])
-		if caseSensitive {
-			output := gogrep.GrepCaseInsensitive(fileContents, pattern)
+	rootPath := flag.Arg(1)
 
-			for i := 0; i < len(output); i++ {
-				fmt.Println(output[i])
-			}
+	paths, err := iofile.ListFiles(rootPath)
+	if err != nil {
+		log.Fatal(err)
+	}
 
-		} else if count {
-			output := gogrep.GrepCount(fileContents, pattern)
-			fmt.Println(output)
+	outputChan := make(chan []string)
+	var wg sync.WaitGroup
+	for _, path := range paths {
+		wg.Add(1)
+		go worker(caseSensitive, count, after, before, path, pattern, outputChan, &wg)
```

Comment:
Why are we checking all the rules inside a goroutine? It is a waste of computation.
If you already know which flag will be used, why not do it outside?

Reply:
Do I create workers for every flag and invoke them by checking the flags within main?

Reply:
Is it possible to use multiple flags at a time?

Reply:
No, Not yet. But I am using both boolVar and Intvar flags, So I need to pass the variables from the Command line to the function as int. I'm almost done with it.

Reply:
Created separate worker functions for each flag which get executed after checking the condition  outside goroutines.